File: apps/desktop/src/platform/browser-host.ts
/**
 * @wfx/app-desktop — the native BrowserHost (R08, productionized R09).
 *
 * The Desktop contained browser surface: a native webview window per
 * surface session (opened by the shell), observed — never steered. R09
 * completes the session lifecycle: the port keeps an adapter-side session
 * registry (`sessions()` — open/navigate/close round-trips are observable)
 * and exposes the honest capability truth (`surfaceCapabilityTruth()`);
 * the EMBED rung engages this SAME port (the contained native webview is
 * where the desktop plays provider embeds — "contained exactly like the
 * browser rung").
 *
 * SECURITY BOUNDARY (frozen remediation architecture, "Media Surface" —
 * the same law the port contract states):
 *
 * - The webview surface is a UX surface, NEVER a circumvention mechanism.
 *   Provider pages stay PROVIDER-OWNED: their markup, scripts, DRM, access
 *   controls, CAPTCHAs, anti-bot controls, rate limits, and geographic
 *   restrictions are never circumvented, intercepted, or modified.
 * - COOKIE/STORAGE ISOLATION IS MANDATORY: every `open()` must carry
 *   `restrictCookies: "isolate"` (the port contract's single legal value).
 *   The shell gives each surface session its own isolated webview data
 *   directory — separate from the WebFlix app window and from every other
 *   surface session — so provider cookies/storage never leak across
 *   sessions or into the app origin. A request with any other mode is
 *   refused BEFORE reaching the shell (typed `invalid-session` refusal,
 *   mirroring the contract's non-optionality).
 * - NAVIGATION IS OBSERVED, NOT BLOCKED, NOT STEERED: the host reports
 *   where the user went (`navigated` events); provider handoffs are honest
 *   observations. The ONLY refusal path is the shell's own policy, which
 *   surfaces as a typed `blocked` event/error with the reason.
 * - NO script injection, NO credential capture, NO content inspection of
 *   provider pages — this port has no mechanism for any of it, by design.
 */

import {
  BrowserSurfaceError,
  type BrowserHostPort,
  type BrowserSurfaceEvent,
  type BrowserSurfaceRequest,
  type BrowserSurfaceSession,
  type Unsubscribe,
} from "@wfx/platform-contracts";

import { isShellIpcError, type ShellIpc, type ShellSurfaceEvent } from "./shell-ipc";

// ---------------------------------------------------------------------------
// R09 — the desktop contained surface's honest CAPABILITY TRUTH
// ---------------------------------------------------------------------------

/**
 * The Desktop contained surface's honest capability truth — what the native
 * webview platform PERMITS. Pure data; never a probe at read time.
 */
export interface DesktopSurfaceCapabilityTruth {
  /** The mount: the shell's native webview window per surface session. */
  readonly mount: "native-webview";
  /** The cookie/storage isolation discipline (per-session data directories). */
  readonly cookieIsolation: "per-session-data-directory";
  /**
   * The honest navigation-observation truth: the SHELL reports where the
   * user went (`navigated` events); provider handoffs are honest
   * observations, never blocked, never steered.
   */
  readonly navigationObservation: "shell-reported";
  /** The Desktop is the full-power reference client — no honest constraint. */
  readonly constrained: false;
  /** The honest constraint note (the unconstrained truth, stated). */
  readonly constraint: string;
}

/**
 * The Desktop contained surface's capability truth (pure; the shell owns
 * the webview, the adapter declares what that truthfully means).
 */
export function desktopSurfaceCapabilityTruth(): DesktopSurfaceCapabilityTruth {
  return {
    mount: "native-webview",
    cookieIsolation: "per-session-data-directory",
    navigationObservation: "shell-reported",
    constrained: false,
    constraint:
      "the desktop contained surface is the full-power reference realization: one isolated native webview per session (its own data directory), navigation observed as the shell reports it",
  };
}

// ---------------------------------------------------------------------------
// The port (R08 + the R09 session-lifecycle productionization)
// ---------------------------------------------------------------------------

/**
 * The Desktop BrowserHostPort: the frozen port plus the R09
 * productionization surface — `sessions()` (the open-session enumeration:
 * open/navigate/close round-trips are observable on it) and the host's own
 * capability truth (`surfaceCapabilityTruth()`).
 */
export interface DesktopBrowserHostPort extends BrowserHostPort {
  /**
   * R09: EVERY open surface session this port has minted (close removes
   * them), in open order, with the last URL the shell reported — the
   * session-lifecycle enumeration. Read-only copies.
   */
  sessions(): readonly { readonly id: string; readonly url: string }[];
  /** R09: the honest capability truth of the desktop contained surface. */
  surfaceCapabilityTruth(): DesktopSurfaceCapabilityTruth;
}

/** Map a shell surface failure onto the typed `BrowserSurfaceError` (1:1 vocabulary). */
function surfaceFailure(operation: string, thrown: unknown): BrowserSurfaceError {
  if (isShellIpcError(thrown)) {
    switch (thrown.code) {
      case "unavailable":
        return new BrowserSurfaceError("unavailable", thrown.detail);
      case "invalid-url":
        return new BrowserSurfaceError("invalid-url", thrown.detail);
      case "blocked":
        return new BrowserSurfaceError("blocked", thrown.detail);
      case "invalid-session":
        return new BrowserSurfaceError("invalid-session", thrown.detail);
      default:
        // Storage/engine-area codes cannot originate from surface commands.
        return new BrowserSurfaceError(
          "unavailable",
          `unexpected shell code '${thrown.code}' during ${operation}: ${thrown.detail}`,
        );
    }
  }
  const detail = thrown instanceof Error ? thrown.message : String(thrown);
  return new BrowserSurfaceError("unavailable", `shell browser host failed during ${operation}: ${detail}`);
}

/** Re-emit one shell surface event to the session's listeners (verbatim mapping). */
function toSurfaceEvent(event: ShellSurfaceEvent): BrowserSurfaceEvent {
  return {
    kind: event.kind,
    sessionId: event.sessionId,
    ...(event.url !== undefined ? { url: event.url } : {}),
    occurredAtMs: event.occurredAtMs,
    ...(event.reason !== undefined ? { reason: event.reason } : {}),
  };
}

/**
 * Build the Desktop `BrowserHostPort` over the shell's webview surface
 * commands. Each `open` mints an isolated surface session handle whose
 * events stream from the shell's surface channel.
 */
export function createShellBrowserHostPort(shell: ShellIpc): DesktopBrowserHostPort {
  /** R09: the adapter-side session registry (the lifecycle enumeration). */
  const sessions = new Map<string, { id: string; url: string; closed: boolean }>();

  return {
    sessions: () =>
      [...sessions.values()]
        .filter((entry) => !entry.closed)
        .map((entry) => ({ id: entry.id, url: entry.url })),

    surfaceCapabilityTruth: () => desktopSurfaceCapabilityTruth(),

    async open(request: BrowserSurfaceRequest): Promise<BrowserSurfaceSession> {
      // The isolation contract is not optional — refuse anything else
      // before the shell is asked (the port's own law).
      if (request.restrictCookies !== "isolate") {
        throw new BrowserSurfaceError(
          "invalid-session",
          `the cookie-isolation contract is not optional (got '${String(request.restrictCookies)}')`,
        );
      }
      let sessionId: string;
      try {
        const opened = await shell.surfaceOpen({
          url: request.url,
          restrictCookies: "isolate",
          purpose: request.purpose,
        });
        sessionId = opened.sessionId;
      } catch (thrown) {
        throw surfaceFailure("open", thrown);
      }

      const listeners = new Set<(event: BrowserSurfaceEvent) => void>();
      let closed = false;

      // R09: the session registry's record (the lifecycle enumeration).
      sessions.set(sessionId, { id: sessionId, url: request.url, closed: false });

      const unsubscribe = shell.onSurfaceEvent((shellEvent) => {
        if (shellEvent.sessionId !== sessionId) return;
        const event = toSurfaceEvent(shellEvent);
        if (event.kind === "closed") {
          closed = true;
          const record = sessions.get(sessionId);
          if (record !== undefined) {
            record.closed = true;
            sessions.delete(sessionId);
          }
        } else if (event.kind === "navigated" && event.url !== undefined) {
          // The enumeration tracks the last URL the shell reported — the
          // observed navigation, never a steered one.
          const record = sessions.get(sessionId);
          if (record !== undefined) record.url = event.url;
        }
        for (const listener of listeners) listener(event);
      });

      return {
        id: sessionId,
        url: request.url,
        async navigate(url: string): Promise<void> {
          try {
            await shell.surfaceNavigate(sessionId, url);
          } catch (thrown) {
            throw surfaceFailure("navigate", thrown);
          }
        },
        async close(): Promise<void> {
          if (closed) {
            unsubscribe();
            return; // closing an already-closed surface is a no-op
          }
          try {
            await shell.surfaceClose(sessionId);
          } catch (thrown) {
            // A failed close stops the event stream regardless — the
            // surface is adapter-best-effort at teardown (documented).
            unsubscribe();
            sessions.delete(sessionId);
            throw surfaceFailure("close", thrown);
          }
          unsubscribe();
          sessions.delete(sessionId);
        },
        subscribe(listener: (event: BrowserSurfaceEvent) => void): Unsubscribe {
          listeners.add(listener);
          return () => {
            listeners.delete(listener);
          };
        },
      };
    },
  };
}

File: apps/desktop/shell/src-tauri/src/browser_host.rs
//! Browser host: the contained webview surface sessions (R08).
//!
//! SECURITY BOUNDARY (frozen remediation architecture, "Media Surface" —
//! the same law the port contract states):
//!
//! - Each `open` creates a CHILD WEBVIEW WINDOW with its OWN isolated
//!   data directory (`{appData}/wfx-desktop/surfaces/<session-id>/`):
//!   the provider's cookies/storage are isolated from the WebFlix app
//!   window AND from every other surface session (the mandatory
//!   `restrictCookies: "isolate"` law — requests carrying any other mode
//!   are refused before a window exists).
//! - Navigation is OBSERVED, never blocked, never steered: URL changes
//!   and window closes emit events to the frontend channel; provider
//!   pages stay provider-owned (no script injection, no credential
//!   capture, no content inspection — by design).
//! - Only http(s) URLs open; anything else rejects with the typed
//!   `invalid-url` failure.

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Mutex;

use tauri::{AppHandle, Emitter, Manager, WebviewWindowBuilder};

use crate::ipc::{now_ms, CommandResult, ShellError, ShellSurfaceEvent, ShellSurfaceHandle, ShellSurfaceOpenRequest};

/// The tracked surface sessions.
pub struct Surfaces {
    app_data_dir: PathBuf,
    sessions: Mutex<HashMap<String, bool>>, // session id -> open
    counter: std::sync::atomic::AtomicU64,
}

impl Surfaces {
    pub fn new(app_data_dir: PathBuf) -> Self {
        Self {
            app_data_dir,
            sessions: Mutex::new(HashMap::new()),
            counter: std::sync::atomic::AtomicU64::new(0),
        }
    }

    fn event(app: &AppHandle, kind: &'static str, session_id: &str, url: Option<String>, reason: Option<String>) {
        let event = ShellSurfaceEvent {
            kind,
            session_id: session_id.to_string(),
            url,
            occurred_at_ms: now_ms(),
            reason,
        };
        let _ = app.emit("wfx://surface", &event);
    }

    /// Open one isolated contained surface at a provider-owned URL.
    pub fn open(&self, app: &AppHandle, request: ShellSurfaceOpenRequest) -> CommandResult<ShellSurfaceHandle> {
        if request.restrict_cookies != "isolate" {
            return Err(ShellError::new(
                "invalid-session",
                "the cookie-isolation contract is not optional (the shell refuses non-isolated surfaces)",
            ));
        }
        let url = tauri::Url::parse(&request.url).map_err(|e| {
            ShellError::new("invalid-url", format!("the URL '{}' is not openable: {}", request.url, e))
        })?;
        if url.scheme() != "http" && url.scheme() != "https" {
            return Err(ShellError::new(
                "invalid-url",
                format!("the contained surface only opens http(s) URLs (got '{}')", url.scheme()),
            ));
        }

        let id = format!("wfx-surface-{}", self.counter.fetch_add(1, std::sync::atomic::Ordering::SeqCst));
        // The isolated data directory for THIS session (cookie/storage
        // isolation — the mandatory law).
        let data_dir = self.app_data_dir.join("wfx-desktop").join("surfaces").join(&id);
        std::fs::create_dir_all(&data_dir)
            .map_err(|e| ShellError::new("io", format!("surface data dir: {e}")))?;

        let builder = WebviewWindowBuilder::new(app, &id, url)
            .title("WebFlix")
            .inner_size(1100.0, 720.0)
            .data_directory(&data_dir);
        builder
            .build()
            .map_err(|e| ShellError::new("unavailable", format!("the webview surface failed to open: {e}")))?;

        if let Ok(mut sessions) = self.sessions.lock() {
            sessions.insert(id.clone(), true);
        }
        Ok(ShellSurfaceHandle { session_id: id })
    }

    /// Navigate a surface (observed, never steered).
    pub fn navigate(&self, app: &AppHandle, session_id: &str, url: &str) -> CommandResult<()> {
        let open = self.sessions.lock().map(|s| *s.get(session_id).unwrap_or(&false)).unwrap_or(false);
        if !open {
            return Err(ShellError::new("invalid-session", format!("no open surface '{session_id}'")));
        }
        let parsed = tauri::Url::parse(url)
            .map_err(|e| ShellError::new("invalid-url", format!("the URL '{url}' is not navigable: {e}")))?;
        if parsed.scheme() != "http" && parsed.scheme() != "https" {
            return Err(ShellError::new("invalid-url", "the surface only navigates to http(s) URLs"));
        }
        if let Some(window) = app.get_webview_window(session_id) {
            window
                .navigate(parsed)
                .map_err(|e| ShellError::new("unavailable", format!("navigate failed: {e}")))?;
        }
        Self::event(app, "navigated", session_id, Some(url.to_string()), None);
        Ok(())
    }

    /// Close a surface exactly once; emits the terminal `closed` event.
    pub fn close(&self, app: &AppHandle, session_id: &str) -> CommandResult<()> {
        let was_open = self
            .sessions
            .lock()
            .map(|mut s| s.remove(session_id).unwrap_or(false))
            .unwrap_or(false);
        if !was_open {
            return Err(ShellError::new("invalid-session", format!("no surface '{session_id}' to close")));
        }
        if let Some(window) = app.get_webview_window(session_id) {
            let _ = window.destroy();
        }
        Self::event(app, "closed", session_id, None, None);
        Ok(())
    }

    /// The window-event bridge: a surface window the USER closed emits
    /// the honest `closed` event (the runtime records an honest stop).
    pub fn on_surface_window_destroyed(app: &AppHandle, label: &str) {
        let surfaces = app.state::<Surfaces>();
        let was_open = surfaces
            .sessions
            .lock()
            .map(|mut s| s.remove(label).unwrap_or(false))
            .unwrap_or(false);
        if was_open {
            Self::event(app, "closed", label, None, None);
        }
    }
}
